vpn/router.go

```diff
@@ -0,0 +1,110 @@
+package vpn
+
+import (
+	"net"
+	"net/netip"
+
+	"tailscale.com/wgengine/router"
+)
+
+func NewRouter(t *Tunnel) router.Router {
+	return &vpnRouter{tunnel: t}
+}
+
+type vpnRouter struct {
+	tunnel *Tunnel
+}
+
+func (*vpnRouter) Up() error {
+	// On macOS, this needs to bring the interface up: `ifconfig <interface> up`.
+	// On Windows, this is a no-op.
+	// How can we/Do we need to initiate this from within the dylib?
+	return nil
+}
+
+func (v *vpnRouter) Set(cfg *router.Config) error {
+	req := convertRouterConfig(cfg)
+	return v.tunnel.ApplyNetworkSettings(v.tunnel.ctx, req)
+}
+
+func (*vpnRouter) Close() error {
+	// There's no cleanup that we need to initiate from within the dylib.
+	return nil
+}
+
+func convertRouterConfig(cfg *router.Config) *NetworkSettingsRequest {
+	v4LocalAddrs := make([]string, 0)
+	v6LocalAddrs := make([]string, 0)
+	for _, addrs := range cfg.LocalAddrs {
+		if addrs.Addr().Is4() {
+			v4LocalAddrs = append(v4LocalAddrs, addrs.String())
+		} else if addrs.Addr().Is6() {
+			v6LocalAddrs = append(v6LocalAddrs, addrs.String())
+		} else {
+			continue
+		}
+	}
+	v4Routes := make([]*NetworkSettingsRequest_IPv4Settings_IPv4Route, 0)
+	v6Routes := make([]*NetworkSettingsRequest_IPv6Settings_IPv6Route, 0)
+	for _, route := range cfg.Routes {
+		if route.Addr().Is4() {
+			//nolint:gosec // outdated rule
+			v4Routes = append(v4Routes, convertToIPV4Route(&route))
+		} else if route.Addr().Is6() {
+			//nolint:gosec // outdated rule
+			v6Routes = append(v6Routes, convertToIPV6Route(&route))
+		} else {
+			continue
+		}
+	}
+	v4ExcludedRoutes := make([]*NetworkSettingsRequest_IPv4Settings_IPv4Route, 0)
+	v6ExcludedRoutes := make([]*NetworkSettingsRequest_IPv6Settings_IPv6Route, 0)
+	for _, route := range cfg.LocalRoutes {
+		if route.Addr().Is4() {
+			//nolint:gosec // outdated rule
+			v4ExcludedRoutes = append(v4ExcludedRoutes, convertToIPV4Route(&route))
+		} else if route.Addr().Is6() {
+			//nolint:gosec // outdated rule
+			v6ExcludedRoutes = append(v6ExcludedRoutes, convertToIPV6Route(&route))
+		} else {
+			continue
+		}
+	}
+
+	return &NetworkSettingsRequest{
+		Mtu: uint32(cfg.NewMTU),
+		Ipv4Settings: &NetworkSettingsRequest_IPv4Settings{
+			Addrs:          v4LocalAddrs,
+			IncludedRoutes: v4Routes,
+			ExcludedRoutes: v4ExcludedRoutes,
+		},
+		Ipv6Settings: &NetworkSettingsRequest_IPv6Settings{
+			Addrs:          v6LocalAddrs,
+			IncludedRoutes: v6Routes,
+			ExcludedRoutes: v6ExcludedRoutes,
+		},
+		TunnelOverheadBytes: 0,  // N/A
+		TunnelRemoteAddress: "", // N/A (?)
```

Comment:
I don't think the tunnel would need this?

Reply:
Yeah, macOS complains if it isn't set to _something_, but it's really meant for cases where there is some single VPN server we are dialing.  You can leave it unset and we can fill it in on the macOS side.